File: python/to_roman.py
#!/usr/bin/env python3

def to_roman(n):
    roman_digits = [
        '',
        'i',
        'ii',
        'iii',
        'iv',
        'v',
        'vi',
        'vii',
        'viii',
        'ix',
        'x'
    ]

    roman_dec = 'IVXLCDM  '

    reversed_n = reversed(str(n))

    ivx = zip(roman_dec[::2], roman_dec[1::2], roman_dec[2::2])

    romans = [
        roman_digits[int(alg)].replace('i', i).replace('v', v).replace('x', x)
        for alg, (i, v, x) in zip(reversed_n, ivx)
    ]

    return ''.join(reversed(romans))


if __name__ == '__main__':
    n = int(input('Enter a number: '))
    print(to_roman(n))

File: haskell/to_roman.hs
romanDigits = [""
              ,"i"
              ,"ii"
              ,"iii"
              ,"iv"
              ,"v"
              ,"vi"
              ,"vii"
              ,"viii"
              ,"ix"
              ,"x"
              ]

romanSymbols = "IVXLCDM  "

toRoman :: Int -> String
toRoman n = foldl1 (++) $ reverse romans
  where n_str  = show n
        rev    = reverse n_str
        algs   = map (read . pure) rev :: [Int]
        algs'  = map (romanDigits !!) algs
        ivx'   = zip3 romanSymbols (drop 1 romanSymbols) (drop 2 romanSymbols)
        ivx    = [x
                 | (i, x) <- zip [0..] ivx'
                 , i `mod` 2 == 0
                 ]
        romans = [[ case c of
                      'i' -> i
                      'v' -> v
                      'x' -> x
                  | c <- alg
                  ]
                 | (alg, (i, v, x)) <- zip algs' ivx
                 ]

main :: IO ()
main = do
  putStrLn "Enter a number: "
  line <- getLine
  let n = read line
  putStrLn . toRoman $ n
